test/test_common/wasm_base.h

```diff
@@ -0,0 +1,125 @@
+#include <cstdio>
+
+#include "envoy/extensions/wasm/v3/wasm.pb.validate.h"
+#include "envoy/server/lifecycle_notifier.h"
+
+#include "common/buffer/buffer_impl.h"
+#include "common/http/message_impl.h"
+#include "common/stats/isolated_store_impl.h"
+#include "common/stream_info/stream_info_impl.h"
+
+#include "extensions/common/wasm/wasm.h"
+#include "extensions/common/wasm/wasm_state.h"
+
+#include "test/mocks/grpc/mocks.h"
+#include "test/mocks/http/mocks.h"
+#include "test/mocks/network/mocks.h"
+#include "test/mocks/server/mocks.h"
+#include "test/mocks/ssl/mocks.h"
+#include "test/mocks/stream_info/mocks.h"
+#include "test/mocks/thread_local/mocks.h"
+#include "test/mocks/upstream/mocks.h"
+#include "test/test_common/environment.h"
+#include "test/test_common/printers.h"
+#include "test/test_common/utility.h"
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Common {
+namespace Wasm {
+
+#define MOCK_CONTEXT_LOG_                                                                          \
+  proxy_wasm::WasmResult log(uint32_t level, absl::string_view message) override {                 \
+    log_(static_cast<spdlog::level::level_enum>(level), message);                                  \
+    return proxy_wasm::WasmResult::Ok;                                                             \
+  }                                                                                                \
+  MOCK_METHOD2(log_, void(spdlog::level::level_enum level, absl::string_view message))
+
+template <typename Base = testing::Test> class WasmTestBase : public Base {
+public:
+  void SetUp() override { clearCodeCacheForTesting(false); }
+
+  void setupBase(const std::string& runtime, const std::string& code,
+                 std::unique_ptr<Context> root_context, std::string root_id = "") {
+    root_context_ = root_context.get();
+    envoy::extensions::wasm::v3::VmConfig vm_config;
+    vm_config.set_vm_id("vm_id");
+    vm_config.set_runtime(absl::StrCat("envoy.wasm.runtime.", runtime));
+    vm_config.mutable_code()->mutable_local()->set_inline_bytes(code);
+    Api::ApiPtr api = Api::createApiForTest(stats_store_);
+    scope_ = Stats::ScopeSharedPtr(stats_store_.createScope("wasm."));
+    auto name = "";
+    auto vm_id = "";
+    auto plugin_configuration = "";
+    plugin_ = std::make_shared<Extensions::Common::Wasm::Plugin>(
+        name, root_id, vm_id, plugin_configuration,
+        envoy::config::core::v3::TrafficDirection::INBOUND, local_info_, &listener_metadata_);
+    // Passes ownership of root_context_.
+    Extensions::Common::Wasm::createWasmForTesting(
+        vm_config, plugin_, scope_, cluster_manager_, init_manager_, dispatcher_, random_, *api,
+        lifecycle_notifier_, remote_data_provider_, std::move(root_context),
+        [this](WasmHandleSharedPtr wasm) { wasm_ = wasm; });
```

Comment:
Hey john, sorry could not look at this earlier.
I needed to add more things here like call configure() and make mocking log optional.
Do you think we should hoist it up here?
I think it makes sense.